index.d.ts

```diff
@@ -562,7 +562,8 @@ declare namespace Moleculer {
 		call<T>(actionName: string): Promise<T>;
 		call<T, P>(actionName: string, params: P, opts?: CallingOptions): Promise<T>;
 
-		mcall<T>(def: Array<MCallDefinition> | { [name: string]: MCallDefinition }, opts?: MCallCallingOptions): Promise<Array<T> | T>;
+		mcall<T>(def: Record<string, MCallDefinition>, opts?: MCallCallingOptions): Promise<Record<string, T>>;
```

Comment:
The `mcall` syntax is really hard to model with TS.  I think this is on the right path (and a likely improvement over what already is there), but there's a potential for issues.

The generic `T` is, more likely than not, going to be incorrect since each called action in the `mcall` is likely to return a different type.  In its normal form, `mcall` is essentially a `Promise.all` and with the `settled` option its essentially a `Promise.allSettled`.

In looking at the built-in TS definitions for `Promise.all`, they aren't really pretty: https://github.com/microsoft/TypeScript/blob/main/lib/lib.es2015.promise.d.ts#L41

They look like they have streamlined things a bit for the `allSettled` definition (probably because it came along later): https://github.com/microsoft/TypeScript/blob/main/lib/lib.es2020.promise.d.ts#L40

Of course, the big challenge here is that the return type of an action can't be inferred the way most promise results can.  Additionally, we must model an overload with `settled` set to `true`.  I'm not really sure how best to approach this in a satisfactory way.  I'll need to think on it a bit.

Overall, this is definitely an improvement over the existing typings in that the object format has _some_ support.  So I think it would be safe to move forward with this, but maybe we should leave a jsdoc on the type with the current limitations.

Is there an advantage to using `mcall` over an array of `Promise.all([ctx.call, ctx.call, ...])`?  We've always used the latter in our environment.

Reply:
Hello,

I totally agree, but in fact, it's the hard think of a function that will return a generic number of generics objects ...


I never use allSettled, but when checking the type of Promise.all they do a manual typing for 10 results ... 

About the rest, the problem with the actual typing, is if I've an easy use case, like each functions return string, I can't directly use it as an array, because it think it can be an object .

about the more complex ways, in general, I retype the object after getting the objects ( to use my personnal types easier ) .

---

I'm too "new in moleculer" to know if there is a difference between .all of call and mcall ... The doc say to use mcall, so I try mcall 😂